### ISSUE TITLE
unbound-control forward fails in captive portal with "failed to prime trust anchor -- could not fetch DNSKEY rrset . DNSKEY IN"

### ISSUE THREAD

**Issue by riastradh** [NLnetLabs/unbound#1005]
**Describe the bug**
I'm on a network with a captive portal whose DHCP server has handed out a resolver, say 10.0.0.1.  This network's resolver server returns NXDOMAIN for `. DNSKEY` queries.

When I try to ask Unbound to temporarily forward all queries to 10.0.0.1, in order to get through the captive portal, it fails all queries with SERVFAIL and logs:

```
Feb  2 16:40:42 singbulli unbound: [15648:0] info: generate keytag query _ta-4f66. NULL IN
Feb  2 16:40:42 singbulli unbound: [15648:0] info: failed to prime trust anchor -- could not fetch DNSKEY rrset . DNSKEY IN
```

**To reproduce**
1. connect to a network with a captive portal that returns NXDOMAIN for `. DNSKEY` queries
2. configure unbound with any combination of:
    - `unbound-control forward 10.0.0.1`
    - `unbound-control forward_add +i . 10.0.0.1`
    - `unbound-control stub_add +i . 10.0.0.1`
       [edited 2024-02-05T21:08Z to correctly reflect the commands I tried; without `.` the forward_and and stub_add commands obviously fail immediately]
3. `unbound-control flush_zone .` and query unbound for example.com A

**Expected behavior**
Unbound returns the answer given by 10.0.0.1.  I don't care if it's true according to DNSSEC; I care about getting access to the internet.

**System:**
 - Unbound version: 1.16.3
 - OS: NetBSD 10.0_RC3
 - `unbound -V` output:
 
```
Version 1.16.3

Configure line: 
Linked libs: pluggable-libevent 2.1.12-stable (it uses kqueue), OpenSSL 3.0.12 24 Oct 2023
Linked modules: dns64 respip validator iterator

BSD licensed, see LICENSE in source package for details.
Report bugs to unbound-bugs@nlnetlabs.nl or https://github.com/NLnetLabs/unbound/issues
```

**Additional information**
I understand dnssec-trigger is supposed to handle this, but (a) my past experience with dnssec-trigger on Ubuntu was that it never once worked for me, and (b) by reading the code, I can't figure out what else it might be doing besides essentially what I already tried with `forward_add +i . <captiveportalresolver>`: https://github.com/NLnetLabs/dnssec-trigger/blob/c8493ef4daaee7a8bed57f6722f92ff39f838937/dnssec-trigger-script.in#L279-L301.  If dnssec-trigger does work in this scenario after all, I would like to know what magic it is doing to convince Unbound to answer these queries without SERVFAIL or REFUSED.

**Comment by wcawijngaards** [comment]
The correct command is to add a forward zone, since it is a resolver that you are asking unbound to send to. Since you want dnssec to be ignored, the '+i' option is nice. That would mean the forward_add line. For both the forward_add line and stub_add line you need to add the zone that is modified, '.', the root zone is what you want here.

But since you already have a trust anchor for the root, the insecure option, as I see reading the code that tries to add it, does nothing. It notices that already a trust anchor exists for the name, and then does not modify the existing entry. So it did not turn off DNSSEC. Because the root trust anchor is configured in config; the '+i' option would work for domain names that fall underneath the configured trust anchor, and thus have no entry yet.

So the forward command is good to have it forward to that resolver, if the network was not captive it would be nicer to go to the authorities on its own without that forward, and then dnssec would work, but since it is captive that is also what dnssec-trigger does.

What dnssec-trigger does for forced insecure, is disable unbound's resolution, temporarily, and set the system to talk to the upstream resolver straight up. What could work for you, is to disable the root trust anchor, by commenting out the 'auto-trust-anchor-file' statement in the config of unbound. That removes the DNSSEC validation from Unbound, and then setting unbound to forward to 10.0.0.1 would likely work.

**Comment by riastradh** [comment]
> For both the forward_add line and stub_add line you need to add the zone that is modified, '.', the root zone is what you want here.

Yes, sorry, that is what I tested -- I transcribed the commands incorrectly into the report above, which I have corrected now.  If I omit the `.` then forward_add and stub_add fail immediately with:

```
error could not find next argument after 10.0.0.1
```

> What could work for you, is to disable the root trust anchor, by commenting out the 'auto-trust-anchor-file' statement in the config of unbound. That removes the DNSSEC validation from Unbound, and then setting unbound to forward to 10.0.0.1 would likely work.

Is there a way to do this with unbound-control?  I'm trying to just temporarily disable DNSSEC validation, and let unbound forward to the captive portal's resolver, without changing any persistent configuration in /etc.

**Comment by wcawijngaards** [comment]
Perhaps put the trust anchor line in a separate text file that is included from the main config. Then change that one line and reload unbound. That would make a sort-of quick change.

Another option could be to turn on permissive mode, but since that does actually perform validation and especially a lot of DNSSEC failure retries because that is not working, it would likely be slow if it works.

With the file change the change would be persistent. With reload_keep_cache the cache is kept while reloading.

**Comment by pemensik** [comment]
The magic in dnssec-trigger is to avoid contacting broken resolvers. Instead use different DNSSEC-aware resolvers. If you need to pass captive portal, it might help to forward only captive portal domain to local resolver. ``forward_add +i onboard-wifi.com <captiveportalresolver>``, where *onboard-wifi.com* is domain used for captive portal. If it responds with NXDOMAIN to ``.`` domain, then it is definitely broken.

If you can send just basic queries to captive portal and other queries to more decent public resolver, it should work better. You would have to otherwise disable validator module in whole unbound to be able to interact with such server. Better to use it only for bare minimum you can. ``+i`` or ``domain-insecure:`` are your friends.
